[quant][graph] Add quantized batch_norm2d op for graph mode

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#36518 [quant][graph] Add quantized batch_norm2d op for graph mode**
* #36494 [quant] Update qbatch_norm name to qbatch_norm2d

Summary:

Test Plan:
test_quantize_script.py test_qbatch_norm

Reviewers:

Subscribers:

Tasks:

Tags:

supriyar committed:
[quant][graph] Add quantized batch_norm2d op for graph mode
Summary:

Test Plan:
test_quantize_script.py test_qbatch_norm

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant][graph] Add quantized batch_norm2d op for graph mode"


Summary:

Test Plan:
test_quantize_script.py test_qbatch_norm

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant][graph] Add quantized batch_norm2d op for graph mode"


Summary:

Test Plan:
test_quantize_script.py test_qbatch_norm

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on test/quantization/test_quantize_script.py:
```diff
@@ -1449,6 +1449,26 @@ def forward(self, x):
                    .run(m.graph)
 
+
+    def test_qbatch_norm(self):
```
can you move this to after `test_quantized_cat`? also can you use `self._test_op_impl` function here?

supriyar committed:
Update on "[quant][graph] Add quantized batch_norm2d op for graph mode"


Summary:

Test Plan:
test_quantize_script.py test_qbatch_norm

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -1009,7 +1019,9 @@ bool InsertObserversHelper::valueNeedsToBeQuantized(Value* v) {
   // Check whether user is quantizable
   for (const auto& use : v->uses()) {
-    if (nodeQuantizable(use.user) || isDynamicLSTMWeight(v, use, is_dynamic)) {
+    if ((nodeQuantizable(use.user) &&
```
&& -> ||?
can you use `valueQuantizable` here?

supriyar replied:
If I make it `||` then I have to remove the op from `_quantizable_call_funcs` else it would insert observers for each input. Removing from `_quantizable_call_funcs` would result in no observers at the output of the node.
I could refactor it a bit more to make the input and output observers for an op explicit and not just depend on `_quantizable_call_funcs`

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -1009,7 +1019,9 @@ bool InsertObserversHelper::valueNeedsToBeQuantized(Value* v) {
-    if (nodeQuantizable(use.user) || isDynamicLSTMWeight(v, use, is_dynamic)) {
+    if ((nodeQuantizable(use.user) &&
+         nodeInputQuantizable(use.user, use.offset)) ||
+        isDynamicLSTMWeight(v, use, is_dynamic)) {
```
I think you can add a `valueQuantizable` function and use that for `isDynamicLSTMWeight` check as well

jerryzh168 replied:
we probably need another refactor to unify these two actually

jerryzh168 replied:
@supriyar could you have a separate PR before this one that refactors `nodeQuantizable` to `valueQuantizable` that takes a `value` as input?
